presto-main/src/test/java/io/prestosql/tests/LogTestDurationListener.java

```diff
@@ -13,30 +13,125 @@
 import org.testng.ITestResult;
 
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.Arrays;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
-
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
 import static io.airlift.units.Duration.nanosSince;
 import static java.lang.String.format;
+import static java.util.stream.Collectors.joining;
 
 public class LogTestDurationListener
-        implements IClassListener, IInvokedMethodListener
+        implements IExecutionListener, IClassListener, IInvokedMethodListener
 {
     private static final Logger LOG = Logger.get(LogTestDurationListener.class);
 
     private static final Duration SINGLE_TEST_LOGGING_THRESHOLD = Duration.valueOf("30s");
     private static final Duration CLASS_LOGGING_THRESHOLD = Duration.valueOf("1m");
+    private static final Duration GLOBAL_IDLE_LOGGING_THRESHOLD = Duration.valueOf("2m");
+
+    private final ScheduledExecutorService scheduledExecutorService;
 
     private final Map<String, TestData> started = new ConcurrentHashMap<>();
+    private final AtomicLong lastChange = new AtomicLong(System.nanoTime());
+    private final AtomicBoolean hangLogged = new AtomicBoolean();
+    private final AtomicBoolean finished = new AtomicBoolean();
+    @GuardedBy("this")
+    private ScheduledFuture<?> monitorHangTask;
+
+    public LogTestDurationListener()
+    {
+        scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(daemonThreadsNamed("TestHangMonitor"));
+    }
+
+    @Override
+    public synchronized void onExecutionStart()
+    {
+        resetHangMonitor();
+        finished.set(false);
+        if (monitorHangTask == null) {
+            monitorHangTask = scheduledExecutorService.scheduleWithFixedDelay(this::checkForTestHang, 10, 10, TimeUnit.SECONDS);
+        }
+    }
+
+    @Override
+    public synchronized void onExecutionFinish()
+    {
+        resetHangMonitor();
+        finished.set(true);
+        // do not stop hang task so notification of hung threads will fire
+        // Note: since the monitor uses daemon threads it will not prevent JVM shutdown
+    }
+
+    private void checkForTestHang()
+    {
+        if (hangLogged.get()) {
+            return;
+        }
+
+        Duration duration = nanosSince(lastChange.get());
+        if (duration.compareTo(GLOBAL_IDLE_LOGGING_THRESHOLD) < 0) {
+            return;
+        }
+
+        if (!hangLogged.compareAndSet(false, true)) {
+            return;
+        }
+
+        Map<String, TestData> runningTests = ImmutableMap.copyOf(started);
+        if (!runningTests.isEmpty()) {
+            String testDetails = runningTests.entrySet().stream()
```

Comment:
instead of logging current tests, you could just log all the threads

- simpler (less state to keep)
- more information (tests may use executors, hand may be in background PrestoServer)

you can do this manually -- or by forcing JVM to do thread dump. JVM-porivded dump would have additional advantage of logging locks, synchronization objects and being compat with tools for thread dump analysis. I don't know the proper API for this (last time i did it, it was outright hack).

Reply:
Yeah, let's do a thread dump after printing the test names. You should be able to do that programmatically via JMX, I think.

Reply:
This is done.  When there are running tests, you get both the test names and the thread dump.